src/PopIII/popiii.cpp

```diff
@@ -0,0 +1,479 @@
+		double vy = renorm_amp * dvy(i, j, k);
+		double const vz = renorm_amp * dvz(i, j, k);
+
+		if (r <= R_sphere) {
+			state.rho = rhotot;
+			state.T = core_temp;
+			eos(eos_input_rt, state);
+
+			// add rotation to vx and vy
+			vx += (-1.0) * distxy * omega_sphere * std::sin(phi);
+			vy += distxy * omega_sphere * std::cos(phi);
+
+		} else {
+			state.rho = 0.01 * rhotot;
+			state.p = 3.595730e-10; // pressure equilibrium - this is the pressure within the core
+			eos(eos_input_rp, state);
+		}
+
+		// call the EOS to set initial internal energy e
+		amrex::Real const e = state.rho * state.e;
+
+		// amrex::Print() << "cell " << i << j << k << " " << state.rho << " " << state.T << " " << e << std::endl;
+
+		state_cc(i, j, k, HydroSystem<PopIII>::density_index) = state.rho;
+		state_cc(i, j, k, HydroSystem<PopIII>::x1Momentum_index) = state.rho * vx;
+		state_cc(i, j, k, HydroSystem<PopIII>::x2Momentum_index) = state.rho * vy;
+		state_cc(i, j, k, HydroSystem<PopIII>::x3Momentum_index) = state.rho * vz;
+		state_cc(i, j, k, HydroSystem<PopIII>::internalEnergy_index) = e;
+
+		Real const Egas = RadSystem<PopIII>::ComputeEgasFromEint(state.rho, state.rho * vx, state.rho * vy, state.rho * vz, e);
+		state_cc(i, j, k, HydroSystem<PopIII>::energy_index) = Egas;
+
+		for (int nn = 0; nn < NumSpec; ++nn) {
+			state_cc(i, j, k, HydroSystem<PopIII>::scalar0_index + nn) = mfracs[nn] * state.rho; // we use partial densities and not mass fractions
+		}
+	});
+}
+
+template <> void RadhydroSimulation<PopIII>::ErrorEst(int lev, amrex::TagBoxArray &tags, amrex::Real /*time*/, int /*ngrow*/)
+{
+	// refine on Jeans length
+	const int N_cells = 64; // inverse of the 'Jeans number' [Truelove et al. (1997)]
+	const amrex::Real G = Gconst_;
+	const amrex::Real dx = geom[lev].CellSizeArray()[0];
+
+	auto const &prob_lo = geom[lev].ProbLoArray();
+	auto const &prob_hi = geom[lev].ProbHiArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+		const int nidx = HydroSystem<PopIII>::density_index;
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			amrex::Real const x = prob_lo[0] + (i + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const y = prob_lo[1] + (j + static_cast<amrex::Real>(0.5)) * dx;
+			amrex::Real const z = prob_lo[2] + (k + static_cast<amrex::Real>(0.5)) * dx;
+
+			Real const rho = state(i, j, k, nidx);
+			Real const pressure = HydroSystem<PopIII>::ComputePressure(state, i, j, k);
+			amrex::GpuArray<Real, Physics_Traits<PopIII>::numMassScalars> massScalars = RadSystem<PopIII>::ComputeMassScalars(state, i, j, k);
+
+			amrex::Real const cs = quokka::EOS<PopIII>::ComputeSoundSpeed(rho, pressure, massScalars);
+
+			const amrex::Real l_Jeans = cs * std::sqrt(M_PI / (G * rho));
+			if (l_Jeans < (N_cells * dx) && rho > 2e-20) {
```

Comment:
Could the density threshold be a parameter instead of hard-coded to `2e-20` here?